File: src/surface.ts
/**
Static facts about a package, derived from its extracted files and manifests
alone — JSON reads, file layout, and line regexes; package code is never parsed
beyond that, and never executed.
`--list` (non-JS registry refs): go/composer/pypi/gem print one import statement
per line in their own syntax; crate:, gh:, and gitlab: have no statically derivable
surface and fall back to the shipped file listing, as does any package whose
lister comes up empty.
`--size`: plain shipped-file sizes — no bundling, no minification, just bytes
on disk — closed by a total.
`ghRepoOf`: the github repository a manifest advertises, for the navigator's `r`
jump — every ecosystem here plus package.json. Fetching delegates to
registries.ts; everything here is reads.
@module
 */
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { basename, dirname, join } from 'node:path';
import { scoped, walkFiles } from './diff.ts';
import { color, csvEnabled, csvRow, paint, stdoutColor } from './env.ts';
import { bad, err, fmtBytes, kb, readJson, sorted } from './public.ts';
import { parseRegistryRef, registryContext } from './registries.ts';

const IDENT = /^[A-Za-z_]\w*$/;
// A missing manifest or directory is the common case here, not an error: every
// lister asks for more spellings than any one package ships.
const readMaybe = (file: string): string => {
  try {
    return readFileSync(file, 'utf8');
  } catch {
    return '';
  }
};
const lsDir = (dir: string): string[] => {
  try {
    return readdirSync(dir);
  } catch {
    return [];
  }
};

// Go packages are directories: every dir with non-test .go files maps to an
// import path under the go.mod module (the ref name when go.mod is absent).
// `internal/`, `testdata/`, `vendor/`, and `_`/`.`-prefixed segments are
// unimportable or ignored by go tooling; `package main` dirs are binaries.
const goSkip = (rel: string): boolean =>
  rel
    .split('/')
    .some(
      (seg) =>
        seg === 'internal' ||
        seg === 'testdata' ||
        seg === 'vendor' ||
        seg.startsWith('.') ||
        seg.startsWith('_')
    );
// No go.mod (pre-modules archives): the ref name is the import path.
const goModule = (pkgDir: string, refName: string): string =>
  /^module\s+(\S+)/m.exec(readMaybe(join(pkgDir, 'go.mod')))?.[1] ?? refName;
const goSurface = (pkgDir: string, refName: string): string[] => {
  const mod = goModule(pkgDir, refName);
  const byDir = new Map<string, string[]>();
  for (const file of walkFiles(pkgDir).keys()) {
    if (!file.endsWith('.go') || file.endsWith('_test.go')) continue;
    const dir = dirname(file);
    const rel = dir === '.' ? '' : dir;
    if (goSkip(rel)) continue;
    const files = byDir.get(rel) ?? [];
    files.push(file);
    byDir.set(rel, files);
  }
  const paths: string[] = [];
  for (const [rel, files] of byDir)
    for (const file of files) {
      const pkg = /^package\s+([A-Za-z_]\w*)/m.exec(readFileSync(join(pkgDir, file), 'utf8'))?.[1];
      if (!pkg) continue;
      if (pkg !== 'main') paths.push(rel ? `${mod}/${rel}` : mod);
      break;
    }
  return sorted(paths);
};

// PSR-4 is a mechanical file→class mapping: namespace prefix + relative path
// with `/`→`\` and `.php` stripped. Only identifier-shaped segments qualify —
// anything else is not autoloadable under PSR-4 anyway.
const composerSurface = (pkgDir: string): string[] => {
  let psr4: Record<string, unknown>;
  try {
    const manifest = readJson<{ autoload?: { 'psr-4'?: Record<string, unknown> } }>(
      join(pkgDir, 'composer.json')
    );
    psr4 = manifest.autoload?.['psr-4'] ?? {};
  } catch {
    return [];
  }
  const classes: string[] = [];
  for (const [ns, target] of Object.entries(psr4)) {
    const dirs = (Array.isArray(target) ? target : [target]).filter(
      (dir): dir is string => typeof dir === 'string'
    );
    for (const dir of dirs) {
      const root = join(pkgDir, dir);
      if (!existsSync(root)) continue;
      for (const file of walkFiles(root).keys()) {
        if (!file.endsWith('.php')) continue;
        const parts = file.slice(0, -4).split('/');
        if (parts.every((part) => IDENT.test(part))) classes.push(ns + parts.join('\\'));
      }
    }
  }
  return sorted(classes);
};

// Wheels and sdist egg-info name their top-level modules outright (src-layout
// sdists keep the egg-info under src/); without that metadata, top-level
// packages are the dirs holding an __init__.py — under src/ when that layout
// exists (its whole point is separating importable code), at the root
// otherwise, minus the conventional non-import trees that also carry
// __init__.py in the wild.
const PY_SKIP = new Set(['doc', 'docs', 'example', 'examples', 'test', 'testing', 'tests']);
const pypiSurface = (pkgDir: string): string[] => {
  const names = new Set<string>();
  for (const base of ['', 'src'])
    for (const ent of lsDir(join(pkgDir, base))) {
      if (!/\.(dist|egg)-info$/.test(ent)) continue;
      // Metadata without top_level.txt reads empty: the layout scan takes over.
      for (const line of readMaybe(join(pkgDir, base, ent, 'top_level.txt')).split('\n'))
        if (IDENT.test(line.trim())) names.add(line.trim());
    }
  if (!names.size)
    for (const base of ['src', '']) {
      for (const ent of lsDir(join(pkgDir, base)))
        if (!PY_SKIP.has(ent) && existsSync(join(pkgDir, base, ent, '__init__.py'))) names.add(ent);
      if (names.size) break;
    }
  return sorted(names);
};

// The lib/ convention is the gem require surface: lib/rails/all.rb answers to
// `require 'rails/all'`. require_paths beyond lib/ live in the gemspec, which
// is Ruby code and stays unread.
const gemSurface = (pkgDir: string): string[] => {
  const lib = join(pkgDir, 'lib');
  if (!existsSync(lib)) return [];
  return sorted(
    [...walkFiles(lib).keys()]
      .filter((file) => file.endsWith('.rb'))
      .map((file) => file.slice(0, -3))
  );
};

// The navigator's `r` target. repository spellings seen in the wild:
// "github:u/r", bare "u/r" shorthand (package.json only), "git+https://….git",
// "git://…", monorepo urls with /tree/… tails, and bare "github.com/u/r".
const GH_REPO = /(?:^github:|github\.com[/:])([\w.-]+)\/([\w.-]+?)(?:\.git)?(?:[/#].*)?$/;
const ghRepoOfUrl = (url: string): string => {
  const hit = GH_REPO.exec(url.trim());
  return hit ? `${hit[1]}/${hit[2]}` : '';
};
// Manifests advertise several urls (homepage, docs, issues, source); the first
// one that reads as a github repository wins, and a package that names none
// simply offers no jump.
const firstGhRepo = (urls: string[]): string => {
  for (const url of urls) {
    const repo = ghRepoOfUrl(url);
    if (repo) return repo;
  }
  return '';
};
const strings = (text: string, re: RegExp): string[] => [...text.matchAll(re)].map((hit) => hit[1]);
// `key = "value"` lines: the toml/cfg shape of every url a manifest declares
// ([package] in Cargo.toml, [project.urls] in pyproject, [metadata] in
// setup.cfg). Array items — git dependencies pinned to a url — never match.
const CONF_URL = /^\s*[\w.\- ]+\s*=\s*["']?([\w+.-]+:\/\/[^"'\s]+)/gm;
const jsRepo = (pkgDir: string): string => {
  let raw: unknown;
  try {
    raw = readJson<{ repository?: unknown }>(join(pkgDir, 'package.json')).repository;
  } catch {
    // No manifest or an unparseable one: nothing advertised.
    return '';
  }
  const url =
    typeof raw === 'string'
      ? raw
      : raw && typeof raw === 'object' && typeof (raw as { url?: unknown }).url === 'string'
        ? (raw as { url: string }).url
        : '';
  // The bare `u/r` shorthand is npm's alone; elsewhere only real urls count.
  return ghRepoOfUrl(url) || (/^[\w.-]+\/[\w.-]+$/.test(url) ? url : '');
};
const crateRepo = (pkgDir: string): string =>
  firstGhRepo(strings(readMaybe(join(pkgDir, 'Cargo.toml')), CONF_URL));
const composerRepo = (pkgDir: string): string => {
  try {
    const manifest = readJson<{ homepage?: unknown; support?: { source?: unknown } }>(
      join(pkgDir, 'composer.json')
    );
    return firstGhRepo(
      [manifest.support?.source, manifest.homepage].filter(
        (url): url is string => typeof url === 'string'
      )
    );
  } catch {
    return '';
  }
};
// Go modules need no manifest field: github-hosted import paths name the
// repository outright, `/v2` major-version suffix and all.
const goRepo = (pkgDir: string, refName: string): string => ghRepoOfUrl(goModule(pkgDir, refName));
// Core metadata is RFC822 — `Home-page: <url>` and `Project-URL: <label>, <url>`
// — in PKG-INFO (sdists), METADATA (wheels), or the egg-info beside them;
// sdists built without it still declare their urls in pyproject/setup.cfg.
const PY_URL = /^(?:Home-page|Project-URL):\s*(?:[^,\n]*,\s*)?(\S+)$/gm;
const pypiRepo = (pkgDir: string): string => {
  const metas = [readMaybe(join(pkgDir, 'PKG-INFO'))];
  for (const base of ['', 'src'])
    for (const ent of lsDir(join(pkgDir, base))) {
      if (!/\.(dist|egg)-info$/.test(ent)) continue;
      metas.push(
        readMaybe(join(pkgDir, base, ent, 'METADATA')),
        readMaybe(join(pkgDir, base, ent, 'PKG-INFO'))
      );
    }
  return (
    firstGhRepo(metas.flatMap((text) => strings(text, PY_URL))) ||
    firstGhRepo(
      ['pyproject.toml', 'setup.cfg'].flatMap((file) =>
        strings(readMaybe(join(pkgDir, file)), CONF_URL)
      )
    )
  );
};
// Gemspecs are Ruby and stay unparsed: the github url quoted in a `homepage` or
// `source_code_uri` assignment is picked out of the text, nothing else is read.
const GEM_URL = /["']([\w+.-]+:\/\/[^"'\s]*github\.com[^"'\s]*)["']/g;
const gemRepo = (pkgDir: string): string =>
  firstGhRepo(
    lsDir(pkgDir)
      .filter((ent) => ent.endsWith('.gemspec'))
      .flatMap((ent) => strings(readMaybe(join(pkgDir, ent)), GEM_URL))
  );
const REPO_OF: Record<string, (pkgDir: string, refName: string) => string> = {
  'composer:': composerRepo,
  'crate:': crateRepo,
  'gem:': gemRepo,
  'go:': goRepo,
  'pypi:': pypiRepo,
};
/** A package's advertised github repo (`owner/name`) and the ecosystem that named it. */
export type GhRepo = { eco: string; repo: string };
const NO_REPO: GhRepo = { eco: '', repo: '' };
export const ghRepoOf = (pkgDir: string, prefix: string = '', refName: string = ''): GhRepo => {
  // gh:/gitlab: extracts are the repository already; there is nowhere to jump.
  if (prefix === 'gh:' || prefix === 'gitlab:') return NO_REPO;
  const known = REPO_OF[prefix];
  if (known) {
    const repo = known(pkgDir, refName);
    return repo ? { eco: prefix.slice(0, -1), repo } : NO_REPO;
  }
  // Prefixless: a JS package or a plain directory of unknown ecosystem, so
  // every manifest gets a look — package.json first, it being the likeliest.
  // Whichever answers names the ecosystem (a rust checkout browsed as plain
  // files is a crate, not an npm package).
  for (const [name, lookup] of [['npm', jsRepo] as const, ...Object.entries(REPO_OF)]) {
    const repo = lookup(pkgDir, refName);
    if (repo) return { eco: name.endsWith(':') ? name.slice(0, -1) : name, repo };
  }
  return NO_REPO;
};

export const surfaceOf = (
  prefix: string,
  name: string,
  pkgDir: string,
  on: boolean = stdoutColor()
): string[] => {
  const wrap = (lines: string[], render: (token: string) => string): string[] =>
    lines.map((token) => render(paint(token, color.cyan, on)));
  const lines =
    prefix === 'go:'
      ? wrap(goSurface(pkgDir, name), (token) => `import "${token}"`)
      : prefix === 'composer:'
        ? wrap(composerSurface(pkgDir), (token) => `use ${token};`)
        : prefix === 'pypi:'
          ? wrap(pypiSurface(pkgDir), (token) => `import ${token}`)
          : prefix === 'gem:'
            ? wrap(gemSurface(pkgDir), (token) => `require '${token}'`)
            : [];
  return lines.length ? lines : sorted(walkFiles(pkgDir).keys());
};

export const registrySurface = async (outDir: string, selector: string): Promise<string[]> => {
  const ref = parseRegistryRef(selector);
  // The import surface is a whole-package fact; a `/path` tail belongs to the
  // file modes (flagless deep link / cat, -s scope).
  if (ref.path) err(`the import surface lists whole packages; drop /${bad(ref.path)}`);
  const got = await registryContext(outDir, ref);
  return surfaceOf(ref.prefix, ref.name, got.pkgDir);
};

// `--size` rows for a registry extract: every shipped file with its byte size,
// path order. The human table closes with a total line; CSV stays rows-only,
// like every other machine listing (sums are one awk away). A `/path` ref tail
// scopes the rows by diff's rule (the file, or a directory's subtree) — and a
// scope matching nothing is a typo'd path, not an empty package; never go silent.
const sizeEntries = (pkgDir: string, sel?: string): [string, number][] => {
  const files = scoped(walkFiles(pkgDir), sel);
  if (sel && !files.size) err(`no shipped file matches /${bad(sel)}; drop the tail to list files`);
  return [...files.entries()].sort(([a], [b]) => a.localeCompare(b));
};
const sizeCsv = (entries: [string, number][]): string[] =>
  entries.map(([path, bytes]) => csvRow([path, `${bytes}b`]));
// The navigator's files-view palette (interactive.ts): directories cyan,
// files unpainted. Rows here are whole relative paths, so the dir prefix
// takes the cyan. No red entry-point pop — a flat listing has no cursor to
// guide toward it.
const paintPath = (path: string, on: boolean): string => {
  const slash = path.lastIndexOf('/');
  const dir = slash >= 0 ? path.slice(0, slash + 1) : '';
  return (dir && paint(dir, color.cyan, on)) + path.slice(slash + 1);
};
const sizeHuman = (entries: [string, number][], on: boolean, archiveBytes?: number): string[] => {
  const total = entries.reduce((sum, [, bytes]) => sum + bytes, 0);
  return [
    ...entries.map(
      ([path, bytes]) => paintPath(path, on) + paint(`  ${kb(bytes)}kb`, color.dim, on)
    ),
    '',
    // Dim like every other summary footer, worded like the -bs one: unpacked ·
    // packed · count.
    paint(
      `${fmtBytes(total)} unpacked` +
        (archiveBytes ? ` · ${fmtBytes(archiveBytes)} packed` : '') +
        ` · ${entries.length} file${entries.length === 1 ? '' : 's'}`,
      color.dim,
      on
    ),
  ];
};
// Headerless like every machine listing; the unit tag keeps rows self-describing.
export const sizesCsv = (pkgDir: string, sel?: string): string[] =>
  sizeCsv(sizeEntries(pkgDir, sel));
export const sizesHuman = (
  pkgDir: string,
  on: boolean = stdoutColor(),
  archiveBytes?: number,
  sel?: string
): string[] => sizeHuman(sizeEntries(pkgDir, sel), on, archiveBytes);
// Degenerate shipped tree for `-s ./input.js`: keep the same row and footer
// grammar without staging a read-only input in a temp directory.
export const fileSizesCsv = (file: string): string[] =>
  sizeCsv([[basename(file), statSync(file).size]]);
export const fileSizesHuman = (file: string, on: boolean = stdoutColor()): string[] =>
  sizeHuman([[basename(file), statSync(file).size]], on);
export const registrySizes = async (outDir: string, selector: string): Promise<string[]> => {
  const ref = parseRegistryRef(selector);
  const got = await registryContext(outDir, ref);
  return csvEnabled()
    ? sizesCsv(got.pkgDir)
    : sizesHuman(got.pkgDir, stdoutColor(), got.archiveBytes);
};

File: bismar-rs/src/surface.rs
//! Static facts: public surface listing and plain file-size stats.
//! Maps bismar's src/surface.ts.

use crate::diff::{scoped, walk_files};
use crate::env::{csv_row, Color};
use crate::public::{fmt_bytes, kb, read_json, resolve_inside, sorted};
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;

static RE_GO_MODULE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?m)^module\s+(\S+)").unwrap());
static RE_GO_PACKAGE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^package\s+([A-Za-z_]\w*)").unwrap());
static RE_IDENT: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[A-Za-z_]\w*$").unwrap());
static RE_DIST_INFO: Lazy<Regex> = Lazy::new(|| Regex::new(r"\.(dist|egg)-info$").unwrap());
static RE_CARGO_NAME: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^name\s*=\s*"([^"]+)""#).unwrap());

// ── Go surface ────────────────────────────────────────────────────────────────

fn go_skip(rel: &str) -> bool {
    rel.split('/').any(|seg| {
        matches!(seg, "internal" | "testdata" | "vendor")
            || seg.starts_with('.')
            || seg.starts_with('_')
    })
}

fn go_module(pkg_dir: &Path, ref_name: &str) -> String {
    let go_mod = pkg_dir.join("go.mod");
    if let Ok(text) = fs::read_to_string(&go_mod) {
        if let Some(cap) = RE_GO_MODULE.captures(&text) {
            return cap[1].to_string();
        }
    }
    ref_name.to_string()
}

pub fn go_surface(pkg_dir: &Path, ref_name: &str) -> Vec<String> {
    let module = go_module(pkg_dir, ref_name);
    let files = walk_files(pkg_dir);
    let mut by_dir: HashMap<String, Vec<String>> = HashMap::new();
    for file in files.keys() {
        if !file.ends_with(".go") || file.ends_with("_test.go") {
            continue;
        }
        let dir = {
            let p = std::path::Path::new(file);
            p.parent()
                .map(|d| d.to_string_lossy().into_owned())
                .unwrap_or_default()
        };
        let rel = if dir == "." || dir.is_empty() {
            String::new()
        } else {
            dir
        };
        if go_skip(&rel) {
            continue;
        }
        by_dir.entry(rel).or_default().push(file.clone());
    }
    let mut paths = Vec::new();
    for (rel, dir_files) in &by_dir {
        for file in dir_files {
            let content = fs::read_to_string(pkg_dir.join(file)).unwrap_or_default();
            if let Some(cap) = RE_GO_PACKAGE.captures(&content) {
                let pkg = &cap[1];
                if pkg != "main" {
                    paths.push(if rel.is_empty() {
                        module.clone()
                    } else {
                        format!("{}/{}", module, rel)
                    });
                    break;
                }
            }
        }
    }
    sorted(paths)
}

// ── Composer (PHP PSR-4) surface ──────────────────────────────────────────────

fn psr4_ident(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_alphanumeric() || c == '_')
}

pub fn composer_surface(pkg_dir: &Path) -> Vec<String> {
    #[derive(serde::Deserialize)]
    struct ComposerJson {
        autoload: Option<AutoloadSection>,
    }
    #[derive(serde::Deserialize)]
    struct AutoloadSection {
        #[serde(rename = "psr-4")]
        psr4: Option<serde_json::Value>,
    }

    let manifest_path = pkg_dir.join("composer.json");
    let Ok(manifest) = read_json::<ComposerJson>(&manifest_path) else {
        return vec![];
    };
    let psr4 = match manifest.autoload.and_then(|a| a.psr4) {
        Some(serde_json::Value::Object(m)) => m,
        _ => return vec![],
    };
    let mut classes = Vec::new();
    for (ns, target) in &psr4 {
        let dirs = match target {
            serde_json::Value::String(s) => vec![s.as_str().to_string()],
            serde_json::Value::Array(arr) => arr
                .iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect(),
            _ => continue,
        };
        for dir in dirs {
            let Some(root) = resolve_inside(pkg_dir, &dir) else {
                continue;
            };
            for file in walk_files(&root).keys() {
                if !file.ends_with(".php") {
                    continue;
                }
                let parts: Vec<&str> = file[..file.len() - 4].split('/').collect();
                if parts.iter().all(|p| psr4_ident(p)) {
                    classes.push(format!("{}{}", ns, parts.join("\\")));
                }
            }
        }
    }
    sorted(classes)
}

// ── PyPI surface ──────────────────────────────────────────────────────────────

static PY_SKIP: &[&str] = &[
    "doc", "docs", "example", "examples", "test", "testing", "tests",
];

pub fn pypi_surface(pkg_dir: &Path) -> Vec<String> {
    let mut names = HashSet::new();
    for base in &["", "src"] {
        let base_dir = if base.is_empty() {
            pkg_dir.to_path_buf()
        } else {
            pkg_dir.join(base)
        };
        if let Ok(entries) = fs::read_dir(&base_dir) {
            for entry in entries.flatten() {
                let name = entry.file_name().to_string_lossy().into_owned();
                if name.ends_with(".dist-info") || name.ends_with(".egg-info") {
                    let top = base_dir.join(&name).join("top_level.txt");
                    if let Ok(content) = fs::read_to_string(&top) {
                        for line in content.lines() {
                            let t = line.trim();
                            if !t.is_empty() && RE_IDENT.is_match(t) {
                                names.insert(t.to_string());
                            }
                        }
                    }
                }
            }
        }
    }
    if names.is_empty() {
        for base in &["src", ""] {
            let base_dir = if base.is_empty() {
                pkg_dir.to_path_buf()
            } else {
                pkg_dir.join(base)
            };
            if let Ok(entries) = fs::read_dir(&base_dir) {
                for entry in entries.flatten() {
                    let name = entry.file_name().to_string_lossy().into_owned();
                    if PY_SKIP.contains(&name.as_str()) {
                        continue;
                    }
                    if base_dir.join(&name).join("__init__.py").exists() {
                        names.insert(name);
                    }
                }
            }
            if !names.is_empty() {
                break;
            }
        }
    }
    sorted(names)
}

// ── Gem surface ───────────────────────────────────────────────────────────────

pub fn gem_surface(pkg_dir: &Path) -> Vec<String> {
    let lib = pkg_dir.join("lib");
    if !lib.exists() {
        return vec![];
    }
    sorted(
        walk_files(&lib)
            .keys()
            .filter(|f| f.ends_with(".rb"))
            .map(|f| f[..f.len() - 3].to_string())
            .collect::<Vec<_>>(),
    )
}

// ── Rust crate surface ────────────────────────────────────────────────────────

pub fn crate_surface(pkg_dir: &Path) -> Vec<String> {
    // Read Cargo.toml for crate name, list pub mod names from src/.
    let cargo = pkg_dir.join("Cargo.toml");
    let name = fs::read_to_string(&cargo)
        .ok()
        .and_then(|t| RE_CARGO_NAME.captures(&t).map(|c| c[1].to_string()))
        .unwrap_or_default();
    if name.is_empty() {
        return vec![];
    }
    sorted(vec![name])
}

// ── Public surface entry point ────────────────────────────────────────────────

/// Generate `import` / `use` lines for a package, or file listing as fallback.
pub fn registry_surface(pkg_dir: &Path, prefix: &str, ref_name: &str, _color: bool) -> Vec<String> {
    let lines = match prefix {
        "go" => go_surface(pkg_dir, ref_name),
        "packagist" => composer_surface(pkg_dir),
        "pypi" => pypi_surface(pkg_dir),
        "gem" => gem_surface(pkg_dir),
        "crate" => crate_surface(pkg_dir),
        _ => vec![],
    };
    if !lines.is_empty() {
        return lines;
    }
    // Fallback: shipped file listing.
    let mut files: Vec<String> = walk_files(pkg_dir).into_keys().collect();
    files.sort();
    files
}

// ── Plain file sizes ──────────────────────────────────────────────────────────

pub fn file_sizes_human(pkg_dir: &Path, sel: Option<&str>, label: &str, color: bool) -> String {
    let files = scoped(walk_files(pkg_dir), sel);
    let mut sorted_files: Vec<(String, u64)> = files.into_iter().collect();
    sorted_files.sort_by(|a, b| a.0.cmp(&b.0));
    let total: u64 = sorted_files.iter().map(|(_, s)| s).sum();
    let col = if color { Color::CYAN } else { "" };
    let reset = if color { Color::RESET } else { "" };
    let mut lines: Vec<String> = sorted_files
        .iter()
        .map(|(path, bytes)| format!("{:>10}  {}", kb(*bytes), path))
        .collect();
    lines.push(format!(
        "\n{}total: {} ({}){}",
        col,
        fmt_bytes(total),
        label,
        reset
    ));
    lines.join("\n")
}

pub fn file_sizes_csv(pkg_dir: &Path, sel: Option<&str>, _label: &str) -> String {
    let files = scoped(walk_files(pkg_dir), sel);
    let mut sorted_files: Vec<(String, u64)> = files.into_iter().collect();
    sorted_files.sort_by(|a, b| a.0.cmp(&b.0));
    let mut lines = vec![csv_row(&["path", "bytes"])];
    for (path, bytes) in &sorted_files {
        lines.push(csv_row(&[path, &bytes.to_string()]));
    }
    lines.join("\n")
}

pub fn sizes_human(entries: &[(String, u64)], label: &str, color: bool) -> String {
    let col = if color { Color::CYAN } else { "" };
    let reset = if color { Color::RESET } else { "" };
    let total: u64 = entries.iter().map(|(_, s)| s).sum();
    let mut lines: Vec<String> = entries
        .iter()
        .map(|(id, bytes)| format!("{:>10}  {}", kb(*bytes), id))
        .collect();
    lines.push(format!(
        "\n{}total: {} ({}){}",
        col,
        fmt_bytes(total),
        label,
        reset
    ));
    lines.join("\n")
}

pub fn sizes_csv(entries: &[(String, u64)]) -> String {
    let mut lines = vec![csv_row(&["id", "bytes"])];
    for (id, bytes) in entries {
        lines.push(csv_row(&[id, &bytes.to_string()]));
    }
    lines.join("\n")
}
